common/src/main/java/com/microsoft/identity/common/internal/cache/MSALOAuth2TokenCache.java

```diff
@@ -1,33 +1,185 @@
+        mAccessTokenSharedPreferences.remove(atCacheItem.getCacheKey());
+    }
+
+    private boolean scopesIntersect(
+            final MsalAccessTokenCacheItem cacheItem1,
+            final MsalAccessTokenCacheItem cacheItem2) {
+        final Set<String> item1Scopes = scopesAsSet(cacheItem1);
+        final Set<String> item2Scopes = scopesAsSet(cacheItem2);
+        for (final String scope : item2Scopes) {
+            if (item1Scopes.contains(scope)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private Set<String> scopesAsSet(final MsalAccessTokenCacheItem cacheItem) {
+        final Set<String> scopeSet = new HashSet<>();
+        final String scopeString = cacheItem.mScope;
+        final String[] scopeArray = scopeString.split("\\s+");
+        scopeSet.addAll(Arrays.asList(scopeArray));
+        return scopeSet;
+    }
+
+    private List<MsalAccessTokenCacheItem> getAllAccessTokensForClientId(final String clientId) {
+        final List<MsalAccessTokenCacheItem> allATs = getAllAccessTokens();
+        final List<MsalAccessTokenCacheItem> allATsForClientId = new ArrayList<>();
+        for (final MsalAccessTokenCacheItem atItem : allATs) {
+            if (clientId.equalsIgnoreCase(atItem.mClientId)) {
+                allATsForClientId.add(atItem);
+            }
+        }
+        return Collections.unmodifiableList(allATsForClientId);
+    }
+
+    private List<MsalAccessTokenCacheItem> getAllAccessTokens() {
+        final Map<String, String> atMap = (Map<String, String>) mAccessTokenSharedPreferences.getAll();
+        final Collection<String> accessTokensAsStrings = atMap.values();
+        final List<MsalAccessTokenCacheItem> atCacheItems = new ArrayList<>();
+
+        for (final String atString : accessTokensAsStrings) {
+            final MsalAccessTokenCacheItem atCacheItem = new Gson().fromJson(atString, MsalAccessTokenCacheItem.class);
+            // TODO clean up this slop...
+            // set up the uniqueIdentifier here...
+            final String rawClientInfo = atCacheItem.mRawClientInfo;
+            String uid = "";
+            String utid = "";
+            final String decodedClientInfo = new String(Base64.decode(rawClientInfo, Base64.URL_SAFE), Charset.forName(StringExtensions.ENCODING_UTF8));
+            final Map<String, String> clientInfoItems;
+            try {
+                clientInfoItems = JsonExtensions.extractJsonObjectIntoMap(decodedClientInfo);
+                uid = clientInfoItems.get("uid"); // TODO where's an appropriate place for these as constants?
+                utid = clientInfoItems.get("utid");
+                atCacheItem.mUserIdentifier =
+                        EncodingUtil.base64UrlEncodeToString(uid)
+                                + "."
+                                + EncodingUtil.base64UrlEncodeToString(utid);
+            } catch (final JSONException e) {
+                // TODO can this happen? I don't think so...
```

Comment:
Theoretically, this Exception shouldn't get thrown, as this data was stash using Gson.toJson().

Figure out what should be done if this happens. Just set to empty String and keep it moving?